### DISCUSSION TITLE
Theming

### DISCUSSION THREAD

**Discussion by jdhitsolutions** [mdgrs-mei/PowerShellRun#59] [General]
I am starting to experiment with theming. One thing that would really help is an annotated layout or some sort of documentation that defines theme properties. For example, what does `CanvasBorder` refer to?

It would also help to be able to resolve settings to human-readable values. For example, if I look at `$o.theme.MarkerForegroundColor` all I see is `PowerShellRun.FontColor`. I have no way of knowing what the actual color is.

**Comment by mdgrs-mei** [comment]
Yes, we should add one image that describes what each theme property means. I'll consider adding that.

Regarding the human-readable display, I haven't looked into it but is it something we can acheive by the PowerShell formating system?

**Reply by jdhitsolutions** [reply]
The theme object has a defined typename, `PowerShellRun.Theme`, so you could create a custom formatting file for it. However, you still need PowerShell commands that can resolve these properties.

![image](https://github.com/user-attachments/assets/eb821c29-cb5d-41cf-acde-67b2c639e69d)

So yes, a custom format file is the right direction. But it needs to be able to resolve a property like `ActionWindowBorderForegroundColor` to a human-readable color. Right now, all I can see is `PowerShellRun.FontColor'. You really need to display the *value*, not the object.

**Reply by mdgrs-mei** [reply]
It turned out that I just needed to override the `ToString` method of `FontColor`.

![image](https://github.com/user-attachments/assets/beff9e1b-ac55-4590-bc12-7ae9ebe7d6c0)

This will be included in the next release.

**Reply by jdhitsolutions** [reply]
I think this will make a big difference.

**Reply by mdgrs-mei** [reply]
v0.10.0 that includes the `ToString()` improvements has been published. Thank you!

**Comment by jdhitsolutions** [comment]
Another idea, which I know is not trivial, would be to revise the theming system to use ANSI escape sequences and $PSStyle.

**Reply by mdgrs-mei** [reply]
I also considered using PSStyle or escape sequences when I implemented the theme. Escape sequences are just strings, so users can set any string to the colors even if they are not color sequences. I chose the robustness and less error handling. It feels a bit late to think about rewriting all.

**Comment by jdhitsolutions** [comment]
I've been able to work out some theming changes.

![image](https://github.com/user-attachments/assets/67bbe88d-524f-4995-8800-a4e294e024a7)

But when I enter a nested entry or a group, I get the default settings.

![image](https://github.com/user-attachments/assets/55eee257-2d43-4d0f-aea8-7d5226a6b07a)

I don't know if this is a bug or if there is a setting I'm not seeing to inherit values.

**Reply by mdgrs-mei** [reply]
Could you show how you set the theme options? I would like to reproduce the issue.

**Reply by jdhitsolutions** [reply]
This code is in my configuration script:

```powershell
$option = [PowerShellRun.SelectorOption]::new()
$option.Prompt = '🤓 '
$option.Theme.Cursor = '▸ '
$option.Theme.CursorForegroundColor = [PowerShellRun.FontColor]::Cyan
$option.Theme.PreviewEnable = $false
$option.Theme.CanvasBorderForegroundColor = [PowerShellRun.FontColor]::Yellow
$roundBorder = [PowerShellRun.BorderSymbol]::new()
$roundBorder.TopLeft = '╭'
$roundBorder.TopRight = '╮'
$roundBorder.BottomLeft = '╰'
$roundBorder.BottomRight = '╯'
$option.Theme.CanvasBorderSymbol = $roundBorder
$option.Theme.PreviewBorderSymbol = $roundBorder
$option.Theme.ActionWindowBorderSymbol = $roundBorder
$option.Theme.ActionWindowBorderForegroundColor = [PowerShellRun.FontColor]::BrightCyan
$option.Theme.SearchBarBorderForegroundColor = [PowerShellRun.FontColor]::BrightGreen

$global:opt = $option
Set-PSRunDefaultSelectorOption $option
```

**Reply by mdgrs-mei** [reply]
Do you use `-Option` parameter for `Invoke-PSRun`? I don't remember why I added the `-Option` parameter. `Set-PSRunDefaultSelectorOption` is enough so I should remove that parameter.

If you pass a different `SelectorOption` instance to `Set-PSRunDefaultSelectorOption` and `Invoke-PSRun`, the issue happens. If that's not the case, it's more complex..

**Reply by jdhitsolutions** [reply]
I am running `Invoke-PSRun` with no parameters.

**Reply by jdhitsolutions** [reply]
I figured the nested theming issue. I added theme changes in a PowerShell session that I started with default theme settings. The ""nested" items were using the original theming not my defined settings. Then I started a new PowerShell session and everything is as I expect. I keep my PowerShell window open for days at a time. I should have thought to restart it.

**Reply by mdgrs-mei** [reply]
Open for days wow.. good hear it worked!

**Comment by jdhitsolutions** [comment]
I like the theme settings now in 0.10.0

```powershell
PS C:\> (Get-PSRunDefaultSelectorOption).theme

CanvasHeightPercentage                      : 60
PreviewSizePercentage                       : 60
NameWidthPercentage                         : 30
PreviewPosition                             : Bottom
CanvasTopMargin                             : 1
CanvasBorderFlags                           : None
CanvasBorderSymbol                          : │ ─ ╭ ╮ ╰ ╯
CanvasBorderForegroundColor                 : Yellow
CanvasBorderBackgroundColor                 :
DefaultForegroundColor                      :
DefaultBackgroundColor                      :
CursorEnable                                : True
Cursor                                      : ▸
Marker                                      : ✓
PromptSymbol                                : >
PromptForegroundColor                       : BrightGreen
```

The next step is to make it easier for the user to set these values. First, the user should be able to use a command like this in the configuration script:

```powershell
$option.Theme.CursorForegroundColor = "Magenta"
```

But this won't work. The user is forced to know the class name.

```powershell
$option.Theme.CursorForegroundColor = [PowerShellRun.FontColor]::Magenta
```

I think related to this is what happens when I try to export theme settings. I have settings defined.

```powershell
PS C:\> (Get-PSRunDefaultSelectorOption).theme | select *foregroundcolor

CanvasBorderForegroundColor                 : Yellow
DefaultForegroundColor                      :
PromptForegroundColor                       : BrightGreen
QueryForegroundColor                        :
SearchBarBorderForegroundColor              : BrightGreen
...
```

I'd like to export the settings so I can import them in a script.

![image](https://github.com/user-attachments/assets/f4ae5bf3-50ef-4962-bf89-1da78183f51a)

But defined values are empty.

Now, cliXML *will* work.

```powershell
(Get-PSRunDefaultSelectorOption).theme | select *foregroundcolor | Export-Clixml -Path d:\temp\p.xml
```

![image](https://github.com/user-attachments/assets/53535d95-983d-47ad-9c7e-27f940710f68)

However, this doesn't necessarily make it easier to import settings. The user should be able to define settings by the enum value without requiring the full name. And JSON exports should work.

**Reply by mdgrs-mei** [reply]
I understand making this syntax possible is helpful sometimes although I prefer writing the class name because it supports tab completion and you can see what color names are available.

```powershell
$option.Theme.CursorForegroundColor = "Magenta"
```

**Reply by mdgrs-mei** [reply]
However, supporting json export will not be trivial to implement. `FontColor` is not a simple Enum (e.g. it supports hex string colors) and it seems that a public property is needed to show up in the json export. That public string property is only needed for the json export, so it requires strong motivation to support json.

Also, how do you handle the json import? I think it might be hard to construct the theme object from json on the user side.

**Comment by jdhitsolutions** [comment]
This is do-able. I exported theme settings to a JSON file and manually set the colors.

```json
{
  "ConsoleCursorShape": 0,
  "KeyRemapModeConsoleCursorShape": 6,
  "CanvasHeightPercentage": 75,
  "PreviewSizePercentage": 60,
  "NameWidthPercentage": 30,
  "PreviewPosition": 1,
  "CanvasTopMargin": 1,
  "CanvasBorderFlags": 0,
  "CanvasBorderSymbol": {
    "Vertical": "│",
    "Horizontal": "─",
    "TopLeft": "╭",
    "TopRight": "╮",
    "BottomLeft": "╰",
    "BottomRight": "╯"
  },
  "CanvasBorderForegroundColor": "Yellow",
  "CursorEnable": true,
  "Cursor": " 👉 ",
  "Marker": "✓ ",
  "PromptSymbol": null,
  "PromptForegroundColor": "BrightGreen",
  "QueryStyle": 0,
  "SearchBarBorderFlags": 8,
  "SearchBarBorderSymbol": {
    "Vertical": "│",
    "Horizontal": "─",
    "TopLeft": "┌",
    "TopRight": "┐",
    "BottomLeft": "└",
    "BottomRight": "┘"
  },
  "SearchBarBorderForegroundColor": "BrightGreen",
  "CursorForegroundColor": "Magenta",
  "MarkerForegroundColor": "Yellow",
  "IconEnable": true,
  "IconForegroundColor": null,
  "NameStyle": 0,
  "NameHighlightForegroundColor": "Cyan",
  "NameHighlightStyle": 0,
  "NameFocusStyle": 4,
  "NameFocusHighlightBackgroundColor": "Magenta",
  "NameFocusHighlightStyle": 4,
  "DescriptionEnable": true,
  "DescriptionForegroundColor": "BrightBlue",
  "DescriptionStyle": 0,
  "DescriptionHighlightForegroundColor": "Cyan",
  "DescriptionHighlightStyle": 0,
  "DescriptionFocusStyle": 4,
  "DescriptionFocusHighlightForegroundColor": "Magenta",
  "DescriptionFocusHighlightStyle": 4,
  "EntryBorderFlags": 0,
  "EntryBorderSymbol": {
    "Vertical": "│",
    "Horizontal": "─",
    "TopLeft": "┌",
    "TopRight": "┐",
    "BottomLeft": "└",
    "BottomRight": "┘"
  },
  "PreviewEnable": true,
  "PreviewStyle": 0,
  "PreviewBorderFlags": 15,
  "PreviewBorderSymbol": {
    "Vertical": "│",
    "Horizontal": "─",
    "TopLeft": "╭",
    "TopRight": "╮",
    "BottomLeft": "╰",
    "BottomRight": "╯"
  },
  "PreviewBorderForegroundColor": "BrightYellow",
  "ActionWindowKeyStyle": 0,
  "ActionWindowKeyFocusStyle": 4,
  "ActionWindowDescriptionForegroundColor": "BrightBlue",
  "ActionWindowDescriptionStyle": 0,
  "ActionWindowDescriptionFocusStyle": 4,
  "ActionWindowBorderEnable": true,
  "ActionWindowBorderSymbol": {
    "Vertical": "│",
    "Horizontal": "─",
    "TopLeft": "╭",
    "TopRight": "╮",
    "BottomLeft": "╰",
    "BottomRight": "╯"
  },
  "ActionWindowBorderForegroundColor": "BrightYellow",
  "ActionWindowScrollBarForegroundColor": "Magenta",
  "TabSize": 6
}
```

In my configuration script, I can import this file and set the theme properties.

```powershell
$option = [PowerShellRun.SelectorOption]::new()

$data = Get-Content C:\Temp\psruntheme.json | ConvertFrom-Json
$option.Prompt = 'Hey Jeff! 🤓 --> '

($option.theme.PSObject.Properties).ForEach({
    if ($_.Name -match "color" -AND $data.$($_.Name)) {
        #Write-Host "Color: $($_.Name) = $($data.$($_.Name))"
        $_.Value = [PowerShellRun.FontColor]::$($data.$($_.Name))
    }
    else {
        $_.Value = $data.$($_.Name)
    }
})
```

I need an extra step to set the color. But I can live with it.

![image](https://github.com/user-attachments/assets/c8bc31a9-9f23-40d1-8734-776ad6d31b67)

**Reply by mdgrs-mei** [reply]
v0.11.0 has been released with the string conversion to color support. You should now be able to remove the color specialization.

```powershell
$option.Theme.CursorForegroundColor = 'Magenta'
```
